modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/logsdb/qa/matchers/source/SourceMatcher.java

```diff
@@ -0,0 +1,171 @@
+
+        var actualMappingAsMap = XContentHelper.convertToMap(BytesReference.bytes(actualMappings), false, actualMappings.contentType())
+            .v2();
+        this.actualNormalizedMapping = MappingTransforms.normalizeMapping(actualMappingAsMap);
+
+        var expectedMappingAsMap = XContentHelper.convertToMap(BytesReference.bytes(expectedMappings), false, actualMappings.contentType())
+            .v2();
+        this.expectedNormalizedMapping = MappingTransforms.normalizeMapping(expectedMappingAsMap);
+
+        this.fieldSpecificMatchers = Map.of("half_float", new FieldSpecificMatcher.HalfFloatMatcher());
+    }
+
+    public MatchResult match() {
+        if (actual.size() != expected.size()) {
+            return MatchResult.noMatch(
+                formatErrorMessage(
+                    actualMappings,
+                    actualSettings,
+                    expectedMappings,
+                    expectedSettings,
+                    "Number of documents does not match, " + prettyPrintLists(actual, expected)
+                )
+            );
+        }
+
+        var sortedAndFlattenedActual = actual.stream()
+            .sorted(Comparator.comparing((Map<String, Object> m) -> parseTimestampToEpochMillis(m.get("@timestamp"))))
+            .map(SourceTransforms::normalize)
+            .toList();
+        var sortedAndFlattenedExpected = expected.stream()
+            .sorted(Comparator.comparing((Map<String, Object> m) -> parseTimestampToEpochMillis(m.get("@timestamp"))))
+            .map(SourceTransforms::normalize)
+            .toList();
+
+        for (int i = 0; i < sortedAndFlattenedActual.size(); i++) {
+            var actual = sortedAndFlattenedActual.get(i);
+            var expected = sortedAndFlattenedExpected.get(i);
+
+            var result = compareSource(actual, expected);
+            if (result.isMatch() == false) {
+                return result;
+            }
+        }
+
+        return MatchResult.match();
+    }
+
+    private MatchResult compareSource(Map<String, List<Object>> actual, Map<String, List<Object>> expected) {
+        for (var expectedFieldEntry : expected.entrySet()) {
+            var name = expectedFieldEntry.getKey();
+
+            var actualValues = actual.get(name);
+            var expectedValues = expectedFieldEntry.getValue();
+
+            MatchResult fieldMatch = matchWithFieldSpecificMatcher(name, actualValues, expectedValues).orElseGet(
+                () -> matchWithGenericMatcher(actualValues, expectedValues)
+            );
+
+            if (fieldMatch.isMatch() == false) {
+                var message = "Source documents don't match for field [" + name + "]: " + fieldMatch.getMessage();
+                return MatchResult.noMatch(message);
+            }
+        }
+
+        return MatchResult.match();
+    }
+
+    private Optional<MatchResult> matchWithFieldSpecificMatcher(String fieldName, List<Object> actualValues, List<Object> expectedValues) {
```

Comment:
Can this method return the specific matcher instead of MatchResult

Reply:
There is no common interface between a field specific matcher and generic matcher. Maybe there should be but i don't immediately see the benefit.